### ISSUE TITLE
ActiveRecord generates invalid SQL string for empty range array condition

### ISSUE THREAD

**Issue by chrisdurheim** [rails/rails#35135]
### Steps to reproduce

When constructing an ActiveRecord statement using an [Array Condition](https://edgeguides.rubyonrails.org/active_record_querying.html#array-conditions), ActiveRecord generates an invalid SQL string when the condition is an empty range (i.e. `0...0`).

As an example: `Post.where('id = ?', 0...0)`

The error raised is `ActiveRecord::StatementInvalid: SQLite3::SQLException: near ")": syntax error`, but I've confirmed this occurs on both SQLite and PostgreSQL.

Here's a project template to reproduce the issue:

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end
end

class Post < ActiveRecord::Base
end

class BugTest < Minitest::Test
  def test_empty_range
    Post.create!

    assert_equal 1, Post.all.count

    # SQL generated is:
    # SELECT "posts".* FROM "posts" WHERE "posts"."id" BETWEEN 0 AND 0
    # This appropriately resolves to an empty ActiveRecord Relation
    assert_equal 0, Post.where(id: 0..0).count

    # SQL generated is:
    # SELECT "posts".* SELECT "posts".* FROM "posts" WHERE "posts"."id" >= 0 AND "posts"."id" < 0
    # This appropriately resolves to an empty ActiveRecord Relation
    assert_equal 0, Post.where(id: 0...0).count

    # SQL generated is
    # SELECT "posts".* FROM "posts" WHERE (id in (0))
    # This appropriately resolves to an empty ActiveRecord Relation
    assert_equal 0, Post.where('id in (?)', 0..0).count

    # SQL generated is
    # SELECT "posts".* FROM "posts" WHERE (id in ())
    # This appropriately resolves to an empty ActiveRecord Relation
    assert_equal 0, Post.where('id in (?)', 0...0).count

    # SQL generated is
    # SELECT "posts".* FROM "posts" WHERE (id = 0)
    # This appropriately resolves to an empty ActiveRecord Relation
    assert_equal 0, Post.where('id = ?', 0..0).count

    # This is the line that throws an ActiveRecord::StatementInvalid error
    # SQL generated is SELECT "posts".* FROM "posts" WHERE (id = )
    # !!! This raises an ActiveRecord::StatementInvalid error !!!
    assert_equal 0, Post.where('id = ?', 0...0).count
  end
end
```
I've noted the behavior for other statements involving ranges, all of which resolve to a valid SQL statement.

### Expected behavior

A where array condition using an empty array should return an empty ActiveRecord relation, as no value can resolve to a range with zero values.

### Actual behavior

ActiveRecord raises an `ActiveRecord::StatementInvalid` error

### System configuration
**Rails version**: 5.2.2

**Ruby version**: 2.5.1

**Comment by chrisdurheim** [comment]
As a follow-up, I dug further into the `ActiveRecord` code and this is where we're getting tripped up:

In `activerecord/lib/active_record/sanitization.rb`:
```ruby
        def quote_bound_value(value, c = connection)
          if value.respond_to?(:map) && !value.acts_like?(:string)
            if value.respond_to?(:empty?) && value.empty?
              c.quote(nil)
            else
              value.map { |v| c.quote(v) }.join(",") # <= This is what's being evaluated
            end
          else
            c.quote(value)
          end
        end
```

In Ruby, Ranges do not respond to `empty?` because in almost every case (other than exclusive end ranges where the start and end value are identical) they aren't ever empty.

By not responding to `empty?`, we're going to the `map` operation, which resolves as an empty array; the `joins` then resolves to an empty string.

I guess there's a bit of a paradigm question here. What *is* an empty exclusive end array where the first and last values are identical.

Is `0...0` essentially `[]` or is it `[0]`? On the one hand, the starting 0 is inclusive, indicating it should be in the list, but the ending 0 indicates it shouldn't be. Right now, it appears that the Ruby interpretation is that the ending zero takes precedence, in which case we essentially have an empty array. If that were the case, I'd expect the outcome for the SQL sanitization and quoting to resolve to `c.quote(nil)` the same as other empty objects.

Any thoughts on this?

**Comment by pjrebsch** [comment]
I'd say that since Ruby implements `(x...x).to_a` as an empty array, the quoted value of an "empty" exclusive range should match that of an empty array, for which the code you referenced above has chosen is equivalent to `c.quote(nil)`.

However, I don't really agree with that as a solution for either case. The returned value from `c.quote(nil)` is `NULL`. Taking your example of `id = ?`, `NULL` could very well be a valid value for `id`, causing the database to return records when one would not expect or desire it.

Having `id` equal to an empty array should mean that no results are returned from the database. Since this can't be guaranteed with any value that's chosen, I propose that the solution be to append `AND 1=0` (like [`ActiveRecord::QueryMethods#none!`](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/relation/query_methods.rb#L741)) to the value returned by `c.quote(nil)` to (better) ensure that the expression as a whole evaluates to `FALSE` in cases where the `value` parameter is an empty set:

```sql
NULL AND 1=0
```

I can't think of any cases right now where doing that might cause a syntax error. The only problem I can think of is if the SQL was comparing a falsey value _and_ the value to be quoted was wrapped in parentheses, but that seems to me like a much rarer occurrence than accidentally matching with `NULL`:

```sql
SELECT 1 WHERE (0 IN (NULL AND 1=0));
SELECT 1 WHERE (0 = (NULL AND 1=0));
```

**Comment by chrisdurheim** [comment]
I agree with that solution. It seems like the real intention here - finding all results with `id` that match a range with no values would be to return an empty `ActiveRecord` relation.

@pjrebsch - do you have a desire to create a PR for this? Otherwise, I can spin one up.

**Comment by pjrebsch** [comment]
@chrisdurheim 
I created a pull request to fix the issue: #35449. 

I chose not to implement it the way that I described above because I think that's a separate discussion and I didn't want it to get in the way of the fix.

**Comment by chrisdurheim** [comment]
@pjrebsch - looks great! Thanks for putting this together!
